### ISSUE TITLE
Object loss in yolov3 vs yolov5

### ISSUE THREAD

**Issue by alvchn** [ultralytics/yolov5#1808]
Firstly, thank you for maintaining this fantastic repository! I am trying to understand some changes in how the _**object loss**_ is calculated compared to earlier yolov3 versions.

**yolov3, earlier version (e.g. from: https://github.com/ultralytics/yolov3/issues/12):**
`loss_obj = BCEWithLogitsLoss1(pred_conf[mask], mask[mask].float())`
`loss_noobj = 0.5 * nM * BCEWithLogitsLoss2(pred_conf[~mask], mask[~mask].float())`

**yolov3 and yolov5, current:**
`BCEobj = nn.BCEWithLogitsLoss(pos_weight=torch.Tensor([h['obj_pw']])).to(device)`
`for i, pi in enumerate(p):`
`...        b, a, gj, gi = indices[i]  # image, anchor, gridy, gridx`
`...        tobj[b, a, gj, gi] = (1.0 - model.gr) + model.gr * iou.detach().clamp(0).type(tobj.dtype)`
`...        loss_obj += BCEobj(pi[..., 4], tobj) * balance[i]`

My questions:
1. For the earlier version, was there any advantage to separately computing _**loss_obj**_, _**loss_noobj**_, and _**mask**_, as opposed to doing it all in one step, as currently done?
2. In the current version, how is the pos/neg imbalance being handled, since _**h['obj_pw'] = 1.0**_ under default settings? Why do we no longer wish to balance positive and negative loss contributions using something like the _**nM**_ multiplier from before?

Asking because I'm noticing (on a custom dataset) lower confidence scores on average with the current repo, and I'm also curious as to the reasoning for why the above change to the loss was made. Thanks!

**Comment by glenn-jocher** [comment]
@alvchn gr stands for giou ratio, and defines the fraction of the objectness loss comprised of iou.

Low confidence scores are indicators of insufficient training, as the confidences also have integrated within them predicted ious, ergo the model is not expecting accurate box predictions for low confidence detections.

h['obj_pw'] = 1.0 is a very interesting hyperparamater, and it's relationship with h['obj'] requires more study ideally, as they have similar but different effects. Theoretically it might be increased in lieu of increasing h['obj'], though this route may also quickly lead to an increase in FPs.

**Comment by Egolee0713** [comment]
I think gr is also a interesting hyper parameter, it may be helps that more accurate box would tend to be higher confident. And gr is default set to be 1.0,  but now I suffer a problem that I have the best f1(calculated by acc and recall) under a very low confident(about 0.3-0.4). It means that there are lots of low confident detections,  as it has many small and crowd objects in my test dateset. I want to know to if there are any methods to help to have best f1 with a higher confident? I have change gr to 0.5 or 0, but helpless.  @glenn-jocher

**Comment by glenn-jocher** [comment]
@yongjingli to specifically optimize F1 you should run hyperparameter evolution with a fitness function weighted towards F1.
